Distribution/Server/Features/Html.hs

```diff
@@ -1519,26 +1603,53 @@ mkHtmlTags HtmlUtilities{..}
       pkgname <- packageInPath dpath
       _       <- lookupPackageName pkgname -- TODO: necessary?
       putTags pkgname
-      return $ toResponse $ Resource.XHtml $ hackagePage "Set tags"
-          [toHtml "Put tags for ", packageNameLink pkgname]
+      currTags <- queryTagsForPackage pkgname
+      revTags <- queryReviewTagsForPackage pkgname
+      let disp = renderReviewTags currTags revTags pkgname
+      return $ toResponse $ Resource.XHtml $ hackagePage "Package Tags" disp
+
+    showPackageTags :: DynamicPath -> ServerPartE Response
+    showPackageTags dpath = do
+      pkgname <- packageInPath dpath
+      currTags <- queryTagsForPackage pkgname
+      revTags <- queryReviewTagsForPackage pkgname
+      let disp = renderReviewTags currTags revTags pkgname
+      return $ toResponse $ Resource.XHtml $ hackagePage "Package Tags" disp
 
     -- serve form for editing, to be received by putTags
     serveTagsForm :: DynamicPath -> ServerPartE Response
     serveTagsForm dpath = do
       pkgname <- packageInPath dpath
       currTags <- queryTagsForPackage pkgname
-      let tagsStr = concat . intersperse ", " . map display . Set.toList $ currTags
-      return $ toResponse $ Resource.XHtml $ hackagePage "Edit package tags"
-        [paragraph << [toHtml "Set tags for ", packageNameLink pkgname],
-         form ! [theclass "box", XHtml.method "post", action $ packageTagsUri tags "" pkgname] <<
-          [ hidden "_method" "PUT"
-          , dlist . ddef . toHtml $ makeInput [thetype "text", value tagsStr] "tags" "Set tags to "
-          , paragraph << input ! [thetype "submit", value "Set tags"]
-          ]]
+      revTags <- queryReviewTagsForPackage pkgname
+      template <- getTemplate templates "tag-edit.html"
+      let toStr = intercalate ", " . map display . Set.toList
+          tagsStr = toStr currTags
+          addns = toStr $ fst  revTags
+          delns = toStr $ snd  revTags
+      trustainer <- authorisedAsMaintainerOrTrustee pkgname
+      user <- authorisedAsAnyUser
```

Comment:
I don't likad adding these new ad-hoc auth functions (and particular the new ones they require in turn), and it's better not to add new auth functions to reduce possible confusion. We can just fail immediately if they're not authorised as any user, so overall we can use:
```
uid <- guardAuthorised [AnyKnownUser]
trustainer <- checkPriviledged uid [InGroup (maintainersGroup pkgname), InGroup trusteesGroup]
```
now ok having said lets not add new auth functions there's a reasonable argument that we should provide a wrapper for `checkPriviledged` in the `Users` feature that does what the underlying `checkPriviledged` does but supplies the `Users` db appropriately (which is what I'm implicitly using above). This at least isn't ad-hoc since it follows the existing pattern.